# 1 Data Preparation

In [38]:
#data handling libs
import numpy as np
import pandas as pd

import os

#sklearn libs
from sklearn.model_selection import train_test_split


In [39]:
loan_data = pd.read_csv("../data/loan_data_2007_2014.csv", dtype={20: str})

In [40]:
loan_data.shape

(466285, 75)

In [41]:

loan_data.head()

Unnamed: 0       id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0           0  1077501    1296599       5000         5000           4975.0   
1           1  1077430    1314167       2500         2500           2500.0   
2           2  1077175    1313524       2400         2400           2400.0   
3           3  1076863    1277178      10000        10000          10000.0   
4           4  1075358    1311748       3000         3000           3000.0   

         term  int_rate  installment grade  ... total_bal_il il_util  \
0   36 months     10.65       162.87     B  ...          NaN     NaN   
1   60 months     15.27        59.83     C  ...          NaN     NaN   
2   36 months     15.96        84.33     C  ...          NaN     NaN   
3   36 months     13.49       339.31     C  ...          NaN     NaN   
4   60 months     12.69        67.79     B  ...          NaN     NaN   

  open_rv_12m open_rv_24m  max_bal_bc all_util total_rev_hi_lim inq_fi  \
0         NaN         NaN         NaN      NaN              NaN    NaN   
1         NaN         NaN         NaN      NaN              NaN    NaN   
2         NaN         NaN         NaN      NaN              NaN    NaN   
3         NaN         NaN         NaN      NaN              NaN    NaN   
4         NaN         NaN         NaN      NaN              NaN    NaN   

  total_cu_tl inq_last_12m  
0         NaN          NaN  
1         NaN          NaN  
2         NaN          NaN  
3         NaN          NaN  
4         NaN          NaN  

[5 rows x 75 columns]

In [42]:
loan_data.tail()

Unnamed: 0       id  member_id  loan_amnt  funded_amnt  \
466280      466280  8598660    1440975      18400        18400   
466281      466281  9684700   11536848      22000        22000   
466282      466282  9584776   11436914      20700        20700   
466283      466283  9604874   11457002       2000         2000   
466284      466284  9199665   11061576      10000        10000   

        funded_amnt_inv        term  int_rate  installment grade  ...  \
466280          18400.0   60 months     14.47       432.64     C  ...   
466281          22000.0   60 months     19.97       582.50     D  ...   
466282          20700.0   60 months     16.99       514.34     D  ...   
466283           2000.0   36 months      7.90        62.59     A  ...   
466284           9975.0   36 months     19.20       367.58     D  ...   

       total_bal_il il_util open_rv_12m open_rv_24m  max_bal_bc all_util  \
466280          NaN     NaN         NaN         NaN         NaN      NaN   
466281          NaN     NaN         NaN         NaN         NaN      NaN   
466282          NaN     NaN         NaN         NaN         NaN      NaN   
466283          NaN     NaN         NaN         NaN         NaN      NaN   
466284          NaN     NaN         NaN         NaN         NaN      NaN   

       total_rev_hi_lim inq_fi total_cu_tl inq_last_12m  
466280          29900.0    NaN         NaN          NaN  
466281          39400.0    NaN         NaN          NaN  
466282          13100.0    NaN         NaN          NaN  
466283          53100.0    NaN         NaN          NaN  
466284          16000.0    NaN         NaN          NaN  

[5 rows x 75 columns]

In [43]:
loan_data.columns.values

array(['Unnamed: 0', 'id', 'member_id', 'loan_amnt', 'funded_amnt',
       'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade',
       'sub_grade', 'emp_title', 'emp_length', 'home_ownership',
       'annual_inc', 'verification_status', 'issue_d', 'loan_status',
       'pymnt_plan', 'url', 'desc', 'purpose', 'title', 'zip_code',
       'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line',
       'inq_last_6mths', 'mths_since_last_delinq',
       'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal',
       'revol_util', 'total_acc', 'initial_list_status', 'out_prncp',
       'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv',
       'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee',
       'recoveries', 'collection_recovery_fee', 'last_pymnt_d',
       'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d',
       'collections_12_mths_ex_med', 'mths_since_last_major_derog',
       'policy_code', 'application_type', 'annual_inc_joint', 'dti_joint',
       'veri

In [44]:
print(loan_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 466285 entries, 0 to 466284
Data columns (total 75 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Unnamed: 0                   466285 non-null  int64  
 1   id                           466285 non-null  int64  
 2   member_id                    466285 non-null  int64  
 3   loan_amnt                    466285 non-null  int64  
 4   funded_amnt                  466285 non-null  int64  
 5   funded_amnt_inv              466285 non-null  float64
 6   term                         466285 non-null  object 
 7   int_rate                     466285 non-null  float64
 8   installment                  466285 non-null  float64
 9   grade                        466285 non-null  object 
 10  sub_grade                    466285 non-null  object 
 11  emp_title                    438697 non-null  object 
 12  emp_length                   445277 non-null  object 
 13 

In [45]:

print(loan_data.isnull().sum())


Unnamed: 0               0
id                       0
member_id                0
loan_amnt                0
funded_amnt              0
                     ...  
all_util            466285
total_rev_hi_lim     70276
inq_fi              466285
total_cu_tl         466285
inq_last_12m        466285
Length: 75, dtype: int64


In [46]:
# Load dataset (selected columns)
selected_columns = [
    'loan_status', 'loan_amnt', 'term', 'int_rate', 'grade', 'sub_grade',
    'emp_length', 'annual_inc', 'home_ownership', 'verification_status',
    'purpose', 'dti', 'delinq_2yrs', 'inq_last_6mths', 'open_acc',
    'pub_rec', 'total_acc'
]

Continuous variables:

Numeric data, can take on any value within a range (e.g., loan amount, income).

Discrete variables:

Countable numeric data (e.g., number of delinquencies).

Categorical variables:

Non-numeric categories (e.g., purpose, grade).

Often converted to Dummy variables or encoded using WoE (Weight of Evidence) for modeling.

Dummy variables:

Binary (0 or 1) variables indicating the presence or absence of categorical features.

Binary target variable (loan_status):

The outcome you're predicting (Default / Non-default).



In [47]:

loan_data = pd.read_csv("../data/loan_data_2007_2014.csv", usecols=selected_columns, low_memory=False)


In [48]:
print(loan_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 466285 entries, 0 to 466284
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   loan_amnt            466285 non-null  int64  
 1   term                 466285 non-null  object 
 2   int_rate             466285 non-null  float64
 3   grade                466285 non-null  object 
 4   sub_grade            466285 non-null  object 
 5   emp_length           445277 non-null  object 
 6   home_ownership       466285 non-null  object 
 7   annual_inc           466281 non-null  float64
 8   verification_status  466285 non-null  object 
 9   loan_status          466285 non-null  object 
 10  purpose              466285 non-null  object 
 11  dti                  466285 non-null  float64
 12  delinq_2yrs          466256 non-null  float64
 13  inq_last_6mths       466256 non-null  float64
 14  open_acc             466256 non-null  float64
 15  pub_rec          

In [49]:
loan_data.isnull().sum()

loan_amnt                  0
term                       0
int_rate                   0
grade                      0
sub_grade                  0
emp_length             21008
home_ownership             0
annual_inc                 4
verification_status        0
loan_status                0
purpose                    0
dti                        0
delinq_2yrs               29
inq_last_6mths            29
open_acc                  29
pub_rec                   29
total_acc                 29
dtype: int64

##  Missing Value Treatment

In [50]:

# emp_length: fill missing with '0 years' (assume no employment history)
loan_data['emp_length'].fillna('0 years', inplace=True)

# delinq_2yrs, inq_last_6mths, open_acc, pub_rec, total_acc: fill missing with 0 (no record)
loan_data[['delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'total_acc']] = \
loan_data[['delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'total_acc']].fillna(0)


## Data Type Conversion

In [51]:
# Convert 'term' from text ('36 months') to numeric (36)
if loan_data['term'].dtype == 'object':
    loan_data['term'] = loan_data['term'].str.extract('(\d+)').astype(int)

# Convert 'int_rate' from percentage string ('10.65%') to float (10.65)
if loan_data['int_rate'].dtype == 'object':
    loan_data['int_rate'] = loan_data['int_rate'].str.rstrip('%').astype(float)


# Convert 'emp_length' to numeric years ('10+ years', '< 1 year' → 10, 0)
loan_data['emp_length'] = loan_data['emp_length'].str.extract('(\d+)').astype(float)

##  Invalid/Unreasonable Values

In [52]:
# Remove rows with annual_inc (annual income) <= 0 (invalid)
loan_data = loan_data[loan_data['annual_inc'] > 0]

# Remove rows with loan_amnt (loan amount) <= 0 (invalid)
loan_data = loan_data[loan_data['loan_amnt'] > 0]

# Keep reasonable debt-to-income ratios (dti) between 0 and 100%
loan_data = loan_data[(loan_data['dti'] >= 0) & (loan_data['dti'] <= 100)]

# Remove duplicates, if any
loan_data.drop_duplicates(inplace=True)


In [53]:
print(loan_data.info())

<class 'pandas.core.frame.DataFrame'>
Index: 466281 entries, 0 to 466284
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   loan_amnt            466281 non-null  int64  
 1   term                 466281 non-null  int64  
 2   int_rate             466281 non-null  float64
 3   grade                466281 non-null  object 
 4   sub_grade            466281 non-null  object 
 5   emp_length           466281 non-null  float64
 6   home_ownership       466281 non-null  object 
 7   annual_inc           466281 non-null  float64
 8   verification_status  466281 non-null  object 
 9   loan_status          466281 non-null  object 
 10  purpose              466281 non-null  object 
 11  dti                  466281 non-null  float64
 12  delinq_2yrs          466281 non-null  float64
 13  inq_last_6mths       466281 non-null  float64
 14  open_acc             466281 non-null  float64
 15  pub_rec              4

In [54]:
print(loan_data.isnull().sum())

loan_amnt              0
term                   0
int_rate               0
grade                  0
sub_grade              0
emp_length             0
home_ownership         0
annual_inc             0
verification_status    0
loan_status            0
purpose                0
dti                    0
delinq_2yrs            0
inq_last_6mths         0
open_acc               0
pub_rec                0
total_acc              0
dtype: int64


In [55]:
loan_data.to_csv('loan_data_clean.csv', index=False)

## 2.PD Model

In [56]:
import scorecardpy as sc

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score, confusion_matrix, recall_score, precision_score, f1_score


In [57]:

# Load the cleaned data
loan_data = pd.read_csv('loan_data_clean.csv')

In [58]:
loan_data['loan_status'].unique()


array(['Fully Paid', 'Charged Off', 'Current', 'Default',
       'Late (31-120 days)', 'In Grace Period', 'Late (16-30 days)',
       'Does not meet the credit policy. Status:Fully Paid',
       'Does not meet the credit policy. Status:Charged Off'],
      dtype=object)

Binary varible/ dummy variable? 

0 (Good / Non-default) ： Fully Paid，Current， Does not meet the credit policy. Status:Fully Paid

1 (Bad / Default)： Charged Off， Default， Late (31-120 days)， In Grace Period， Does not meet the credit policy. Status:Charged Off

In [59]:
# Binary encode loan_status clearly
good_status = ['Fully Paid', 'Current', 'Does not meet the credit policy. Status:Fully Paid']

loan_data['loan_status'] = loan_data['loan_status'].apply(lambda x: 0 if x in good_status else 1)

# Check results after conversion
print(loan_data['loan_status'].value_counts())


loan_status
0    410949
1     55332
Name: count, dtype: int64


In [60]:
# Split the data into train and test sets (e.g., 80% train / 20% test)
train, test = train_test_split(loan_data, test_size=0.2, random_state=42)


Why Chooese 80% train 20% test?  overfit  vs underfit

In [61]:
# WoE binning based on training data

bins = sc.woebin(train, y='loan_status')

[INFO] creating woe binning ...


/Users/annieyi/anaconda3/lib/python3.11/site-packages/scorecardpy/condition_fun.py:40: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  datetime_cols = dat.apply(pd.to_numeric,errors='ignore').select_dtypes(object).apply(pd.to_datetime,errors='ignore').select_dtypes('datetime64').columns.tolist()
/Users/annieyi/anaconda3/lib/python3.11/site-packages/scorecardpy/condition_fun.py:40: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  datetime_cols = dat.apply(pd.to_numeric,errors='ignore').select_dtypes(object).apply(pd.to_datetime,errors='ignore').select_dtypes('datetime64').columns.tolist()
/Users/annieyi/anaconda3/lib/python3.11/site-packages/scorecardpy/condition_fun.py:40: UserWarning: Could not infer format, so eac

WOE = ln (Distibution of Good)/ (Distribution of Bad) 

Positive WOE means Distribution of Goods > Distribution of Bads

Negative WOE means Distribution of Goods < Distribution of Bads

Hint : Log of a number > 1 means positive value. If less than 1, it means negative value.  

In [ ]:
# WoE Transformation (linearity assumption) #Improve Model Stability (overfit)  #Capture Non-linear Relationships # Handle Missing and Outliers #Enhance Interpretability
train_woe = sc.woebin_ply(train, bins)
test_woe = sc.woebin_ply(test, bins)


[INFO] converting into woe values ...
[INFO] converting into woe values ...


WoE transformation in logistic regression is primarily to handle categorical variables effectively and capture potential non-linear relationships

**Information Value (IV)**

Information value is a useful technique to select important variables in a predictive model. It represents how much independent information it brings originally to explain dependent value. It helps to rank variables on the basis of their importance. The IV is calculated using the following formula :  

IV = ∑ (% of non-events - % of events) * WOE

IV categories  
* Less than 0.02 => 	Not useful for prediction   
* 0.02 to 0.1	=> Weak predictive Power  
* 0.1 to 0.3	=> Medium predictive Power  
* 0.3 to 0.5 => 	Strong predictive Power  
* '>0.5 =>	Suspicious Predictive Power  


In [63]:
# IV Calculation
iv_df = sc.iv(train_woe, y='loan_status')
print(iv_df.sort_values('info_value', ascending=False))

                   variable  info_value
9              int_rate_woe    0.320908
14                grade_woe    0.294599
7             sub_grade_woe    0.293527
3        inq_last_6mths_woe    0.045506
2            annual_inc_woe    0.044108
5                  term_woe    0.041400
12              purpose_woe    0.028524
11                  dti_woe    0.025541
15  verification_status_woe    0.020303
4        home_ownership_woe    0.019947
6            emp_length_woe    0.005665
13            total_acc_woe    0.005622
1             loan_amnt_woe    0.005109
0              open_acc_woe    0.000375
8           delinq_2yrs_woe    0.000354
10              pub_rec_woe    0.000251


In [64]:
# Select features with IV > 0.1
selected_features = iv_df[iv_df['info_value'] > 0.1]['variable'].tolist()
print("Selected features (IV > 0.1):", selected_features)


Selected features (IV > 0.1): ['int_rate_woe', 'grade_woe', 'sub_grade_woe']


In [65]:
# Final datasets for PD modeling
train_final = train_woe[['loan_status'] + selected_features]
test_final = test_woe[['loan_status'] + selected_features]

In [66]:
# Save datasets for modeling
train_final.to_csv('train_final.csv', index=False)
test_final.to_csv('test_final.csv', index=False)

## Model Evaluation

In [67]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score, confusion_matrix, recall_score, precision_score, f1_score
import pandas as pd

In [68]:
# Load datasets
train = pd.read_csv('train_final.csv')
test = pd.read_csv('test_final.csv')

X_train, y_train = train.drop('loan_status', axis=1), train['loan_status']
X_test, y_test = test.drop('loan_status', axis=1), test['loan_status']


In [75]:
# Train Logistic Regression model
model = LogisticRegression(max_iter=500)
model.fit(X_train, y_train)

LogisticRegression(max_iter=500)

In [ ]:
# Predict default probabilities (PD)
y_pred_prob = model.predict_proba(X_test)[:, 1]

# Evaluate AUC(Area Under the ROC Curve)
auc = roc_auc_score(y_test, y_pred_prob)
print(f"AUC (no threshold applied): {auc:.3f}")


AUC (no threshold applied): 0.652


AUC of 0.65 means my model can correctly rank defaulting borrowers higher than non-defaulting borrowers 65% of the time.


Real-life credit risk scenarios, it's very common and acceptable for PD models to achieve an AUC between 0.70 to 0.80.
An AUC above 0.80 is considered excellent and not very common, while values beyond 0.85 are rare and could indicate potential data leakage or unrealistic features."

In [72]:
# Choose threshold (0.2 based on business requirement)
threshold = 0.2
y_pred = (y_pred_prob >= threshold).astype(int)


In [77]:
# Compute metrics at chosen threshold
recall = recall_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)
tn, fp, fn, tp = cm.ravel()
specificity = tn / (tn + fp)

# Print detailed metrics
print(f"Recall: {recall:.3f}")
print(f"Precision: {precision:.3f}")
print(f"Specificity: {specificity:.3f}")
print(f"F1 Score: {f1:.3f}")

Recall: 0.206
Precision: 0.224
Specificity: 0.904
F1 Score: 0.214


Recall ： 0.206  - Out of all actual defaulted borrowers, my model successfully identified approximately 20.6% of them

Percision: 0.224 - Among the borrowers predicted by my model to default, about 22.4% actually did default

Specificity: 0.904 - My model correctly identified approximately 90.4% of borrowers who did not default

F1 Score: The F1 score, which balances Precision and Recall, is 21.4%. This moderate score indicates there is significant room to enhance the model, especially in correctly predicting actual defaults.


EL : Estimating expected loss

EAD: Exposure at defalut

LGD: Loss Given Defalut(%)

EL = PD X EAD X LGD

In [ ]:
# Original test data for EAD (loan_amnt)
test_original = pd.read_csv('loan_data_clean.csv').iloc[X_test.index]

# Assume LGD = 50%  #
LGD = 0.5

# Calculate EL
test_original['PD'] = y_pred_prob
test_original['EL'] = test_original['PD'] * test_original['loan_amnt'] * LGD

# Show sample results
print(test_original[['loan_amnt', 'PD', 'EL']])


       loan_amnt        PD           EL
0           5000  0.196650   491.623830
1           2500  0.128762   160.952792
2           2400  0.219986   263.983379
3          10000  0.142961   714.806314
4           3000  0.028894    43.341123
...          ...       ...          ...
93252      10100  0.083049   419.398965
93253      12000  0.083049   498.295800
93254      20000  0.045123   451.225263
93255      18000  0.175635  1580.716063
93256      14400  0.196650  1415.876632

[93257 rows x 3 columns]
